### ISSUE TITLE
pixel aspect ratio

### ISSUE THREAD

**Issue by alex9311** [xyonix/vannot#42]
as we discussed over email, we are having issues when videos have non-square pixel aspect ratios. You mentioned that you are not able to get that data from the video file. 

If we pass the pixel aspect ratio to vannot the same way we currently pass the frame height/width, could you use that information to generate valid frames upon export?

**Comment by clint-tseng** [comment]
pooooossibly. but it would be a lot of really weird difficult math because the shape data will have been coded against a somewhat arbitrarily incorrect projection to begin with.

it would be far easier to just synthesize the correct data in the first place. i think it's likely possible that given a PAR at the same time a width/height is given the correct canvas projection can be calculated in the first place. but in this case, all vannot would be doing is multiplying the width by the given PAR and praying that it works out correctly (having spent a while googling this issue now, i'm unclear on exactly what the specification thinks should be done with the PAR to obtain the display resolution). if you already have eg ffmpeg analyzing the file to learn its storage resolution, it may be easier to just ask it what the actual display resolution is.

**Comment by alex9311** [comment]
I see, yeah we have all the data ffprobe gives about the videos. So  I see one video that results in the bug has  
```
    Stream #0:0(und): Video: h264 (High) (avc1 / 0x31637661), yuv420p, 854x480 [SAR 320:427 DAR 4:3], 1038 kb/s, 29.94 fps, 30 tbr, 15360 tbn, 60 tbc (default)
```

so you're saying if we pass you the DAR that would help?

**Comment by clint-tseng** [comment]
the reason i'm really hesitant to say "yes" here (or to make it something vannot tries to solve) is that i have not been able, despite some effort, to actually determine what the specification says you are supposed to do with the storage resolution and the PAR or DAR to actually get the display resolution. absent more-specific rules, if i have a box and you tell me to make it some other aspect ratio, there are an infinite number of answers i could possibly give back to you which would satisfy your constraints.

my suspicion is that the number of rows always stays the same and the number of columns change to meet the ratio requirement. but i'm really not certain about it. so my hope was that the toolchain you had would actually somehow tell you what _it_ thinks the final display _resolution_ should be, because it will have a far greater understanding about these things than i do.

**Comment by alex9311** [comment]
hey clint, thanks for the response.

I found this function in fluent-ffmpeg (node wrapper for ffmpeg) called [keepDar](https://github.com/fluent-ffmpeg/node-fluent-ffmpeg/blob/master/README.md#keepdar-force-keeping-display-aspect-ratio). It is used on video to convert to square pixels. The method sets the pixel aspect ratio (aka SAR or PAR) to 1 and adjusts the heightxwidth of the video. The exact math it does (in ffmpeg speak) is:
```
-filter:v scale=w='if(gt(sar,1),iw*sar,iw)':h='if(lt(sar,1),ih/sar,ih)',setsar=1
```
which effectively means "if the PAR is greater than one, multiply the frame width by the PAR. If the PAR is less than one, divide the frame height by the PAR. If I apply that math to the `video.height` and `video.width` I pass in on videos with non-square pixels, it appears to solve our bug of misaligned polygons on the exported jpg. 

This seems a bit too easy to be true...

**Comment by clint-tseng** [comment]
ha. i guess that makes some amount of sense. do you want that logic worked into vannot, or are you fine with implementing it as a preprocessing step?

**Comment by alex9311** [comment]
I would lean towards having the login in vannot, since this would likely be a problem for other people. Sound OK? We could make it an optional float  `video.pixel_aspect_ratio`?

**Comment by clint-tseng** [comment]
okay—two commits have been pushed to master.

one draws a grey bounding box around the area vannot believes to be the canvas area. if the grey box does not match the video area, it is a sign that the given parameters are incorrect.
the second accepts `video.pixelAspectRatio` as prescribed above.

the README has been updated with these details as well.

**Comment by alex9311** [comment]
thanks Clint! I'm seeing an issue on the bounding box with non-1 PAR. I left comments on the commits about it

**Comment by alex9311** [comment]
Great! I've tested with PAR 1, 8:9, and 4:3.

The only issue I ran into was that the width of the exported frame when the naive width is 640 and the PAR is 4:3 is not an integer (853.333) which upset the python script trying to read the jpg. I rounded out the jpg frame dimensions and it worked fine.

Unless non-integer dimensions on jpgs is a big no-no in general, I dont mind dealing with that on our end
